# Question 9:
Write a program to read the data from the following link, perform data analysis and answer the following questions:

Link - https://data.wa.gov/api/views/f6w7-q2d2/rows.csv?accessType=DOWNLOAD

Insights to be drawn:
* Get all the cars and their types that do not qualify for clean alternative fuel vehicle
* Get all TESLA cars with the model year, and model type made in Bothell City.
* Get all the cars that have an electric range of more than 100, and were made after 2015
* Draw plots to show the distribution between city and electric vehicle type


In [23]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import cufflinks as cf
import plotly.offline as pyo
pyo.init_notebook_mode(connected=True)
cf.go_offline()

from warnings import filterwarnings
filterwarnings('ignore')

# Reading Data

In [10]:
url = 'https://data.wa.gov/api/views/f6w7-q2d2/rows.csv?accessType=DOWNLOAD'
df = pd.read_csv(url, encoding = 'utf-8')
df.head()

VIN (1-10)    County     City State  Postal Code  Model Year    Make  \
0  5YJXCAE26J    Yakima   Yakima    WA      98908.0        2018   TESLA   
1  JHMZC5F37M    Kitsap  Poulsbo    WA      98370.0        2021   HONDA   
2  5YJ3E1EB0K      King  Seattle    WA      98199.0        2019   TESLA   
3  1N4AZ0CP5D      King  Seattle    WA      98119.0        2013  NISSAN   
4  5YJSA1E21H  Thurston    Lacey    WA      98516.0        2017   TESLA   

     Model                   Electric Vehicle Type  \
0  MODEL X          Battery Electric Vehicle (BEV)   
1  CLARITY  Plug-in Hybrid Electric Vehicle (PHEV)   
2  MODEL 3          Battery Electric Vehicle (BEV)   
3     LEAF          Battery Electric Vehicle (BEV)   
4  MODEL S          Battery Electric Vehicle (BEV)   

  Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
0           Clean Alternative Fuel Vehicle Eligible             238   
1           Clean Alternative Fuel Vehicle Eligible              47   
2           Clean Alternative Fuel Vehicle Eligible             220   
3           Clean Alternative Fuel Vehicle Eligible              75   
4           Clean Alternative Fuel Vehicle Eligible             210   

   Base MSRP  Legislative District  DOL Vehicle ID  \
0          0                  14.0       141151601   
1          0                  23.0       171566447   
2          0                  36.0         9426525   
3          0                  36.0       211807760   
4          0                  22.0       185810306   

              Vehicle Location                              Electric Utility  \
0  POINT (-120.56916 46.58514)                                    PACIFICORP   
1  POINT (-122.64681 47.73689)                        PUGET SOUND ENERGY INC   
2  POINT (-122.40092 47.65908)  CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)   
3   POINT (-122.3684 47.64586)  CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)   
4  POINT (-122.75379 47.06316)                        PUGET SOUND ENERGY INC   

   2020 Census Tract  
0       5.307700e+10  
1       5.303509e+10  
2       5.303301e+10  
3       5.303301e+10  
4       5.306701e+10

In [11]:
df.dtypes

VIN (1-10)                                            object
County                                                object
City                                                  object
State                                                 object
Postal Code                                          float64
Model Year                                             int64
Make                                                  object
Model                                                 object
Electric Vehicle Type                                 object
Clean Alternative Fuel Vehicle (CAFV) Eligibility     object
Electric Range                                         int64
Base MSRP                                              int64
Legislative District                                 float64
DOL Vehicle ID                                         int64
Vehicle Location                                      object
Electric Utility                                      object
2020 Census Tract       

In [12]:
df.isna().sum()

VIN (1-10)                                             0
County                                                 3
City                                                   3
State                                                  0
Postal Code                                            3
Model Year                                             0
Make                                                   0
Model                                                222
Electric Vehicle Type                                  0
Clean Alternative Fuel Vehicle (CAFV) Eligibility      0
Electric Range                                         0
Base MSRP                                              0
Legislative District                                 305
DOL Vehicle ID                                         0
Vehicle Location                                      33
Electric Utility                                       3
2020 Census Tract                                      3
dtype: int64

In [13]:
df.columns

Index(['VIN (1-10)', 'County', 'City', 'State', 'Postal Code', 'Model Year',
       'Make', 'Model', 'Electric Vehicle Type',
       'Clean Alternative Fuel Vehicle (CAFV) Eligibility', 'Electric Range',
       'Base MSRP', 'Legislative District', 'DOL Vehicle ID',
       'Vehicle Location', 'Electric Utility', '2020 Census Tract'],
      dtype='object')

# Data Analysis

## Insight 1:
Get all the cars and their types that do not qualify for clean alternative fuel vehicle

**Getting all the unique values in CAFV Eligibility**

In [14]:
df['Clean Alternative Fuel Vehicle (CAFV) Eligibility'].unique()

array(['Clean Alternative Fuel Vehicle Eligible',
       'Not eligible due to low battery range',
       'Eligibility unknown as battery range has not been researched'],
      dtype=object)

### Data

In [18]:
df_not_eligible_for_CAFV = df[df['Clean Alternative Fuel Vehicle (CAFV) Eligibility'] != 'Clean Alternative Fuel Vehicle Eligible']
df_not_eligible_for_CAFV.head(3)

VIN (1-10)    County       City State  Postal Code  Model Year  Make  \
6  3FA6P0PU1G  Thurston   Tumwater    WA      98501.0        2016  FORD   
7  WA1F2AFY8P  Thurston      Lacey    WA      98516.0        2023  AUDI   
8  1FADP5CU0E  Thurston  Rochester    WA      98579.0        2014  FORD   

    Model                   Electric Vehicle Type  \
6  FUSION  Plug-in Hybrid Electric Vehicle (PHEV)   
7    Q5 E  Plug-in Hybrid Electric Vehicle (PHEV)   
8   C-MAX  Plug-in Hybrid Electric Vehicle (PHEV)   

  Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
6             Not eligible due to low battery range              19   
7             Not eligible due to low battery range              23   
8             Not eligible due to low battery range              19   

   Base MSRP  Legislative District  DOL Vehicle ID  \
6          0                  22.0       347508836   
7          0                  22.0       227213871   
8          0                  20.0       229811151   

              Vehicle Location        Electric Utility  2020 Census Tract  
6  POINT (-122.89166 47.03956)  PUGET SOUND ENERGY INC       5.306701e+10  
7  POINT (-122.75379 47.06316)  PUGET SOUND ENERGY INC       5.306701e+10  
8  POINT (-123.08743 46.82175)  PUGET SOUND ENERGY INC       5.306701e+10

### Count

In [17]:
df_not_eligible_for_CAFV.shape

(69892, 17)

Total **69892 Vehicles** out of the given data are inelligible for CAFV

### Type of Vehicles that are not eligible for CAFV

In [28]:
df_not_eligible_for_CAFV['Electric Vehicle Type'].value_counts().iplot(kind='bar', title='Vehicle type of CAFV-inelligible vehicles')

* Hover over the bar graph to get the actual count

## Insight 2:
Get all TESLA cars with the model year, and model type made in Bothell City.

### Data

In [38]:
df_tesla_bothell = df [(df.City == 'Bothell') & (df.Make == 'TESLA')][['Make', 'City', 'Model Year', 'Model']]
df_tesla_bothell.head(3)

Make     City  Model Year    Model
39   TESLA  Bothell        2020  MODEL 3
217  TESLA  Bothell        2022  MODEL 3
218  TESLA  Bothell        2022  MODEL Y

### Count

In [39]:
df_tesla_bothell.shape

(2407, 4)

There are a total of **2407 Tesla Vehicles** manufactured in Bothell

### Model Type

In [40]:
df_tesla_bothell['Model'].value_counts().iplot(kind='bar', title='Model Type of Bothell manufactured Tesla Vehicles')

* Hover over the bar graph to get the actual count

### Model Year

In [41]:
df_tesla_bothell['Model Year'].value_counts().iplot(kind='bar', title='Year of manufacture of Tesla Vehicles from Bothell')

* Hover over the bar graph to get the actual count

## Insight 3:
Get all the cars that have an electric range of more than 100, and were made after 2015

In [50]:
df_2015_plus_cars_with_electric_range_more_than_100 = df[(df['Electric Range']>100) & (df['Model Year'] > 2015)]
df_2015_plus_cars_with_electric_range_more_than_100.head(3)

VIN (1-10)    County     City State  Postal Code  Model Year   Make  \
0  5YJXCAE26J    Yakima   Yakima    WA      98908.0        2018  TESLA   
2  5YJ3E1EB0K      King  Seattle    WA      98199.0        2019  TESLA   
4  5YJSA1E21H  Thurston    Lacey    WA      98516.0        2017  TESLA   

     Model           Electric Vehicle Type  \
0  MODEL X  Battery Electric Vehicle (BEV)   
2  MODEL 3  Battery Electric Vehicle (BEV)   
4  MODEL S  Battery Electric Vehicle (BEV)   

  Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
0           Clean Alternative Fuel Vehicle Eligible             238   
2           Clean Alternative Fuel Vehicle Eligible             220   
4           Clean Alternative Fuel Vehicle Eligible             210   

   Base MSRP  Legislative District  DOL Vehicle ID  \
0          0                  14.0       141151601   
2          0                  36.0         9426525   
4          0                  22.0       185810306   

              Vehicle Location                              Electric Utility  \
0  POINT (-120.56916 46.58514)                                    PACIFICORP   
2  POINT (-122.40092 47.65908)  CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)   
4  POINT (-122.75379 47.06316)                        PUGET SOUND ENERGY INC   

   2020 Census Tract  
0       5.307700e+10  
2       5.303301e+10  
4       5.306701e+10

### Count

In [51]:
df_2015_plus_cars_with_electric_range_more_than_100.shape

(34066, 17)

There are **34066 Cars** with electric range more than 100 and manufactured after 2015

## Insight 4:
Draw plots to show the distribution between city and electric vehicle type

### Citywise number of BEVs and PHEVs

In [60]:
stacked_bar = df.groupby(['City', 'Electric Vehicle Type']).size().unstack().fillna(0)
stacked_bar.iplot(kind='bar', barmode='stack', title='Stacked Bar Chart: City vs Electric Vehicle Type (Absolute Numbers)')

**Note:**
    The above chart shows (in absolute numbers) the **type of EVs in each city**. We can compare the absolute number of BEVs and PHEVs in all cities and see the proportional preference to both these types *within the city*. However we cannot compare the proportional preference of BEV vs PHEVs among two or more *different cities* due to it being in interval scale. To achieve this we compare the normalized value_counts of each city in the next chart.

### Citywise proportional preference of BEVs vs PHEVs

In [61]:
proportions = df.groupby('City')['Electric Vehicle Type'].value_counts(normalize=True).unstack().fillna(0)
proportions = proportions.sort_values('Battery Electric Vehicle (BEV)')
proportions.iplot(kind='bar', barmode='stack', title='Stacked Bar Chart: City vs Electric Vehicle Types (Proportion)')

* Zoom in to see details of each city. Hover to see actual ratio